### ISSUE TITLE
Turn a Table Collection to a Tree Sequence in C

### ISSUE THREAD

**Issue by dmliustat** [tskit-dev/tskit#403]
I'm currently writing a C program sampling genealogies given a sample of gene sequences. In my C program, I use the C API tskit to store the ancestral history and save the tree sequences in .trees files. Then I hope to load these .trees files for further analysis in Python.

I notice that the C API allows to save the Table Collection object in .trees file, but when I load these .trees file in Python using tskit, I got error saying "Key not found". I understand that there should be a one-to-one correspondence between the tables and the .trees files. So I'm wondering is there a way to load the Table Collections (which are also with the .trees suffix) in Python and then convert them to Tree Sequences?

I also tried to convert the Table Collections to Tree Sequences in my C program. Unfortunately, I didn't find much description about the functions regarding Tree Sequences. To be specific, I used tsk_treeseq_init() to initialize a tsk_treeseq_t object but it failed. Is the second argument, tsk_table_collection_t *tables, the table collection that I want the tree sequence to store? Or should I use tsk_treeseq_copy_tables()  instead (actually I tried but got some errors)? Could you please provide some guidance on these functions?

To summarize, I have generated a Table Collection in my C program and viewed it in txt file. I don't see obvious errors in the table collection. My problem is that the Table Collection cannot be read into Python. Could you please help me convert a Table Collection to a Tree Sequence, either in Python or the C language?

**Comment by jeromekelleher** [comment]
Hi @dmliustat, welcome to tskit-dev!

Firstly, apologies for the poor documentation on the C API. It's great to see someone else use the C API, so please do raise issues when you find problems. 

You certainly *should* be able to write out a table collection to file and read it back from Python. This is what the [forward simulation](https://tskit.readthedocs.io/en/latest/c-api.html#basic-forwards-simulator) example does. If you use ``tsk_table_collection_dump`` it should write out a valid set of tables. When you dump to file, can you read it back in using the C API (using [this example](https://tskit.readthedocs.io/en/latest/c-api.html#tree-iteration), e.g.?

Can you give some more details about the errors that you're getting when you try to load the file into Python?

**Comment by dmliustat** [comment]
Hi Dr. Kelleher, thank you so much for your quick reply!

I did use tsk_table_collection_dump for saving the ancestral history, but when I read the .trees files into python, I got the folllowing error:

---------------------------------------------------------------------------
FileFormatError                           Traceback (most recent call last)
~\Anaconda\lib\site-packages\tskit\trees.py in load(path)
   1523     try:
-> 1524         return TreeSequence.load(path)
   1525     except exceptions.FileFormatError as e:

~\Anaconda\lib\site-packages\tskit\trees.py in load(cls, path)
   2029         ts = _tskit.TreeSequence()
-> 2030         ts.load(str(path))
   2031         return TreeSequence(ts)

FileFormatError: Key not found

**Comment by dmliustat** [comment]
More details about this error...
---------------------------------------------------------------------------
The above exception was the direct cause of the following exception:

FileFormatError                           Traceback (most recent call last)
<ipython-input-2-bfb73345d3a3> in <module>
----> 1 simTree = tskit.load("C:/Users/blabl/Dropbox/output/out_0.trees")

~\Anaconda\lib\site-packages\tskit\trees.py in load(path)
   1524         return TreeSequence.load(path)
   1525     except exceptions.FileFormatError as e:
-> 1526         formats.raise_hdf5_format_error(path, e)
   1527 
   1528 

~\Anaconda\lib\site-packages\tskit\formats.py in raise_hdf5_format_error(filename, original_exception)
    257     except (IOError, OSError, KeyError):
    258         # We want to keep a useful error message here as well as the chaining history.
--> 259         raise exceptions.FileFormatError(str(original_exception)) from original_exception
    260 
    261 

FileFormatError: Key not found

**Comment by dmliustat** [comment]
Hi Dr. Kelleher,
I can't load my .trees files using the C API, either. The error was "Exception thrown: read access violation. tree_sequence->tables was nullptr." and it occurs to this line of code:
num_nodes = tree_sequence->tables->nodes.num_rows;
under the function tsk_tree_init (trees.c).
I used ### ### tsk_table_collection_print_state for viewing and I'm attaching the txt file. Hope it helps.
[out.txt](https://github.com/tskit-dev/tskit/files/3850500/out.txt)

**Comment by jeromekelleher** [comment]
Hmm, interesting. Would you mind attaching the tree sequence file here (or emailing it to me) please?

Also, what version of the C API are you using?

It looks like you're developing this on Windows, is that correct? That's excellent (I'm super happy if people are using this on Windows!), but there may be some wrinkles we haven't encountered because we don't use Windows much in the core developers (yet).

**Comment by dmliustat** [comment]
Yes, I'm developing on Windows and using C_0.99.2. To use the C API, I did a small change to the source code (though may not be the reason of failure):

On trees.c Line 745, I got a building error "Error C4703 potentially uninitialized local pointer variable 'row' used", so I changed `uint32_t *restrict row` to `uint32_t *restrict row = NULL` on line 637.
[trees.zip](https://github.com/tskit-dev/tskit/files/3853196/trees.zip)

**Comment by jeromekelleher** [comment]
Thanks for this @dmliustat. It's strange: when you look at the file using the underlying [kastore](kastore.readthedocs.io/en/latest/)  it contains no keys at all:
```
$ kastore ls out.trees
```

I don't understand how this could happen.

Would you mind making a minimal reproducible example for the C API please? That is, if you could replicate the problem that you have here with a short C program, then hopefully we'll be able to get to the bottom of it quickly.

> On trees.c Line 745, I got a building error "Error C4703 potentially uninitialized local pointer variable 'row' used", so I changed uint32_t *restrict row to uint32_t *restrict row = NULL on line 637.
trees.zip

I've opened another issue to track this @dmliustat. If you'd like to open a PR with your fix we'd really appreciate it.

**Comment by dmliustat** [comment]
Looks like it's not because I'm developing on Windows. My program samples the ancestral tree with randomly chosen genetic events and this "no key" problem doesn't always occur. I've attached two trees, with and without error, respectively. Do you have any idea what could cause it? I'll also try to make a reproducible example.
[tree.zip](https://github.com/tskit-dev/tskit/files/3873561/tree.zip)

**Comment by jeromekelleher** [comment]
The only thing I can think of is that an error of some sort is occuring, and it's not being caught. Would you mind pasting in the code chunk where you're dumping the tables to file please?

**Comment by dmliustat** [comment]
Thank you so much, Dr. Kelleher. Yes, you are right. It is an error of sorting. The requirement "edges must be sorted first by child ID" was not met. The error message of missing key made me omit the error of sorting. I think this issue is solved.

**Comment by jeromekelleher** [comment]
Ah, that's good to hear. I still don't understand how an empty file (the missing key error) was written out though, and I just want to make sure there isn't a bug here---if this is happening without an error being set somewhere, it's definitely a bug. Would you mind outlining how the error happened (or just paste in some code) please?

**Comment by dmliustat** [comment]
I've pasted some code below that can reproduce the error. Actually `tsk_table_collection_dump`  fails to dump the tree when the requirement "edges must be sorted first by child ID" was not met, and thus there is no key in the .trees file. 

---------------------------------------------------------------------------------------------------------
#include <stdlib.h>
#include <stdio.h>

#include "tables.h"


#define _CRT_SECURE_NO_DEPRECATE

int main()
{
	int ret;
	int node, edge, mut, site;
	short i;
	tsk_table_collection_t tables;
	FILE* test;
	char filename[200];

	ret = tsk_table_collection_init(&tables, 0);
	for (i = 0; i < 4; i++) {
		node = tsk_node_table_add_row(&tables.nodes, 1, 0.0, TSK_NULL, TSK_NULL, NULL, 0);
	}

	tables.sequence_length = 1.0;

	site = tsk_site_table_add_row(&tables.sites, 0.0, 0, 0, NULL, 0);
	site = tsk_site_table_add_row(&tables.sites, 0.25, 0, 0, NULL, 0);
	site = tsk_site_table_add_row(&tables.sites, 0.33, 0, 0, NULL, 0);
	site = tsk_site_table_add_row(&tables.sites, 0.66, 0, 0, NULL, 0);
	site = tsk_site_table_add_row(&tables.sites, 1.0, 0, 0, NULL, 0);


	/*first event: coalescence */
	node = tsk_node_table_add_row(&tables.nodes, 0, 1.0, TSK_NULL, TSK_NULL, NULL, 0);
	/*first child 1, then child 0; this should incur the error*/
	edge = tsk_edge_table_add_row(&tables.edges, 0.0, 1.0, node, 1);
	edge = tsk_edge_table_add_row(&tables.edges, 0.0, 1.0, node, 0);


	/*second event: recombination */
	node = tsk_node_table_add_row(&tables.nodes, 0, 2.0, TSK_NULL, TSK_NULL, NULL, 0);
	edge = tsk_edge_table_add_row(&tables.edges, 0.66, 1.0, node, 2);
	node = tsk_node_table_add_row(&tables.nodes, 0, 2.0, TSK_NULL, TSK_NULL, NULL, 0);
	edge = tsk_edge_table_add_row(&tables.edges, 0.0, 0.66, node, 2);

	/*third event: mutation */
	mut = tsk_mutation_table_add_row(&tables.mutations, 0.33, 1, -1, 0, 0, NULL, 0);

	snprintf(filename, 200, "output_test.trees");
	ret = tsk_table_collection_dump(&tables, filename, 0);
	if (ret < 0) fprintf(stderr, "Dumping trees failed.\n");
	
	snprintf(filename, 200, "output_test.txt");
	fopen_s(&test, filename, "w");
	tsk_table_collection_print_state(&tables, test);
	fclose(test);

	ret = tsk_table_collection_clear(&tables);

	return 0;

}

**Comment by jeromekelleher** [comment]
Ah, I see, thanks for this example, it's very helpful. I think it is a bug that we write out an empty file when there is an error condition like this, so I've opened an issue to track this (#423).

I know it's tedious, but every return value really does need to be checked, as this is how tskit communicates error conditions. For example, each of your ``add_rows`` above can potentially throw an error, so you should check, .e.g., ``if (node < 0) { printf("Error: %s", tsk_strerror(node); exit(1);} ``. There's a discussion on this in the docs [here](https://tskit.readthedocs.io/en/latest/c-api.html#error-handling).
